kernel/section.ml

```diff
@@ -0,0 +1,248 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *   INRIA, CNRS and contributors - Copyright 1999-2019       *)
+(* <O___,, *       (see CREDITS file for the list of authors)           *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+open Util
+open Names
+open Univ
+
+module NamedDecl = Context.Named.Declaration
+
+type _ section_kind =
+| SecMono : [ `mono ] section_kind
+| SecPoly : [ `poly ] section_kind
+
+type _ section_universes =
+| SecMonoUniv : [ `mono ] section_universes
+| SecPolyUniv : Name.t array * UContext.t -> [ `poly ] section_universes
+
+type section_entry =
+| SecDefinition of Constant.t
+| SecInductive of MutInd.t
+
+type 'a entry_map = 'a Cmap.t * 'a Mindmap.t
+
+type 'a section = {
+  sec_context : int;
+  (** Length of the named context suffix that has been introduced locally *)
+  sec_universes : 'a section_universes;
+  (** Universes local to the section *)
+  sec_entries : section_entry list;
+  (** Definitions introduced in the section *)
+  sec_data : 'a section_universes entry_map;
+}
+
+(** Sections can be nested with the proviso that no monomorphic section can be
+    opened inside a polymorphic one. The reverse is allowed. *)
+type t = {
+  sec_poly : [ `poly ] section list;
+  sec_mono : [ `mono ] section list;
+}
+
```

Comment:
That's certainly a weird data structure for a stack, I personally find the logic in `close_section` very convoluted and prone to errors. I tho understand the purpose, but IMHO it could be worth to explore:
```
type t = section_wrap * section_wrap list
and section_wrap =
 | Mono :: [ `mono ] section -> section_wrap
 | Poly :: [ `poly] section -> section_wrap
```

Reply:
The current implementation statically enforces that no monomorphic section can appear polymorphic section though...

Reply:
>The current implementation statically enforces that no monomorphic section can appear inside a polymorphic section though...

We only really need to prevent monomorphic definitions/inductives from depending on polymorphic section universes. Maybe we shouldn't structure the code around this stronger section nesting invariant.

Reply:
Indeed after writing my comment I was close to delete it as the current implementation has some advantages, on the other hand I was thinking more on the lines of what @SkySkimmer points out.

Anyways consider the comment as a very soft suggestion, the code is OK as-is in that case if you want [tho it complicates the implementation of non-empty sections by typing a bit]